paths_cli/compiling/core.py

```diff
@@ -0,0 +1,300 @@
+        return attr
+
+    def __call__(self, *args, **kwargs):
+        # check correct call signature here?
+        return self.loader(*args, **kwargs)
+
+    def to_json_schema(self, schema_context=None):
+        dct = {
+            'type': self.json_type,
+            'description': self.description,
+        }
+        return self.name, dct
+
+
+class Builder:
+    """Builder is a wrapper class to simplify writing builder functions.
+
+    When the compiled parameters dictionary matches the kwargs for your
+    class, you can create a valid delayed builder function with
+
+    .. code::
+
+        builder = Builder('import_path.for_the.ClassToBuild')
+
+    Additionally, this class provides hooks for functions that run before or
+    after the main builder function. This allows many objects to be built by
+    implementing simple functions and hooking themn together with Builder.
+
+    Parameters
+    ----------
+    builder : Union[Callable, str]
+        primary callable to build an object, or string representing the
+        fully-qualified path to a callable
+    remapper : Callable[[Dict], Dict], optional
+        callable to remap the the mapping of ???
+    after_build : Callable[[Any, Dict], Any], optional
+        ccallable to update the created object with any additional
+        information from the original dictionary.
+    """
+    def __init__(self, builder, *, remapper=None, after_build=None):
+        if remapper is None:
+            remapper = lambda dct: dct
+        if after_build is None:
+            after_build = lambda obj, dct: obj
+        self.remapper = remapper
+        self.after_build = after_build
+        self.builder = builder
+
+    def __call__(self, **dct):
+        # TODO: change this InstanceCompilerPlugin.build to make this better
+        if isinstance(self.builder, str):
+            module, _, func = self.builder.rpartition('.')
+            builder = import_thing(module, func)
+        else:
+            builder = self.builder
+
+        ops_dct = self.remapper(dct.copy())
+        obj = builder(**ops_dct)
+        after = self.after_build(obj, dct.copy())
+        return after
+
+
+class InstanceCompilerPlugin(OPSPlugin):
+    """
+
+    Parameters
+    ----------
+    builder : Callable
+        Function that actually creates an instance of this object. Note that
+        the :class:`.Builder` class is often a useful tool here (but any
+        callable is allowed).
+    parameters : List[:class:`.Parameter]
+        Descriptions of the paramters for that the builder takes.
+    name : str
+        Name used in the text input for this object.
+    aliases : List[str]
+        Other names that can be used.
+    requires_ops : Tuple[int, int]
+        version of OpenPathSampling required for this functionality
+    requires_cli : Tuple[int, int]
+        version of the OPS CLI requires for this functionality
+    """
+    SCHEMA = "http://openpathsampling.org/schemas/sim-setup/draft01.json"
```

Comment:
just in case I forget while looking at the test, we should test to make sure this stays reachable

Reply:
Full JSON schema implementation coming later. This URL *doesn't* exist yet. It (or something like it) will be used to distinguish the core functionality from external plugins (and to generate documentation).